src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -0,0 +1,118 @@
+            return ReadResponse(responseStream);
+        }
+
+        private async Task SendRequest(HttpRequest request, Stream requestStream)
+        {
+            using (var writer = new StreamWriter(requestStream, Encoding.ASCII, BufferSize, leaveOpen: true))
+            {
+                await _datadogHttpHeaderHelper.WriteLeadingHeaders(request, writer, requestStream).ConfigureAwait(false);
+
+                foreach (var header in request.Headers)
+                {
+                    await _datadogHttpHeaderHelper.WriteHeader(writer, header).ConfigureAwait(false);
+                }
+
+                await _datadogHttpHeaderHelper.WriteEndOfHeaders(writer).ConfigureAwait(false);
+            }
+
+            await request.Content.CopyToAsync(requestStream).ConfigureAwait(false);
+            Logger.Debug("Datadog HTTP: Flushing stream.");
+            await requestStream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private async Task<HttpResponse> ReadResponse(Stream responseStream)
+        {
+            var headers = new HttpHeaders();
+            int statusCode = 0;
+            string responseMessage = null;
+
+            // hack: buffer the entire response so we can seek
+            var memoryStream = new MemoryStream();
+            responseStream.CopyTo(memoryStream);
+            memoryStream.Position = 0;
+            int streamPosition = 0;
+
+            using (var reader = new StreamReader(memoryStream, Encoding.ASCII, detectEncodingFromByteOrderMarks: false, BufferSize, leaveOpen: true))
+            {
+                // HTTP/1.1 200 OK
+                // HTTP/1.1 XXX MESSAGE
+                string line = reader.ReadLine();
+                streamPosition += reader.CurrentEncoding.GetByteCount(line) + DatadogHttpHeaderHelper.CrLfLength;
+
+                const int statusCodeStart = 9;
+                const int statusCodeLength = 3;
+                const int startOfMessage = 13;
+
+                if (!int.TryParse(line.Substring(statusCodeStart, statusCodeLength), out statusCode))
+                {
+                    throw new DatadogHttpRequestException("Invalid response, can't parse status code. Line was:" + line);
+                }
+
+                responseMessage = line.Substring(startOfMessage);
+
+                // read headers
+                while (true)
+                {
+                    line = await reader.ReadLineAsync().ConfigureAwait(false);
+                    streamPosition += reader.CurrentEncoding.GetByteCount(line) + DatadogHttpHeaderHelper.CrLfLength;
+
+                    if (line == string.Empty)
+                    {
+                        // end of headers
+                        break;
+                    }
+
+                    var headerParts = line.Split(':');
+
+                    if (headerParts.Length != 2)
+                    {
+                        Logger.Warning("Malformed header: {0}", line);
```

Comment:
Are the headers we absolutely NEED? Should we skip out if there are any malformed headers?

Reply:
Found this triggers on Date headers like `Date: Fri, 11 Dec 2020 18:18:42 GMT`

Reply:
Refactored to parse byte by byte after the leading line as colons are valid in the value per the spec.